test/Sema/ConversionConstraintTest.cpp

```diff
@@ -0,0 +1,343 @@
+    
+    // But should be allowed explicitly
+    EXPECT_TRUE(testExplicitConversion(int64Type, int32PtrType));
+}
+
+// Test enum conversions
+TEST_F(ConversionConstraintTest, EnumToInteger)
+{
+    // Enum to integer should NOT be allowed implicitly
+    EXPECT_FALSE(testImplicitConversion(enumType, int32Type));
+    
+    // But should be allowed explicitly
+    EXPECT_TRUE(testExplicitConversion(enumType, int32Type));
+}
+
+TEST_F(ConversionConstraintTest, IntegerToEnum)
+{
+    // Integer to enum should NOT be allowed implicitly
+    EXPECT_FALSE(testImplicitConversion(int32Type, enumType));
+    
+    // But should be allowed explicitly
+    EXPECT_TRUE(testExplicitConversion(int32Type, enumType));
+}
+
+// Test type variable conversions
+TEST_F(ConversionConstraintTest, TypeVariableConversions)
+{
+    // Type variables should always allow conversion
+    EXPECT_TRUE(testImplicitConversion(typeVar, int32Type));
+    EXPECT_TRUE(testImplicitConversion(int32Type, typeVar));
+    EXPECT_TRUE(testImplicitConversion(typeVar, typeVar));
+    
+    // Same for explicit conversions
+    EXPECT_TRUE(testExplicitConversion(typeVar, int32Type));
+    EXPECT_TRUE(testExplicitConversion(int32Type, typeVar));
+    EXPECT_TRUE(testExplicitConversion(typeVar, typeVar));
+}
+
+// Test same type conversions
+TEST_F(ConversionConstraintTest, SameTypeConversions)
+{
+    // Same types should always be valid
+    EXPECT_TRUE(testImplicitConversion(int32Type, int32Type));
+    EXPECT_TRUE(testImplicitConversion(float32Type, float32Type));
+    EXPECT_TRUE(testImplicitConversion(int32PtrType, int32PtrType));
+    EXPECT_TRUE(testImplicitConversion(boolType, boolType));
+    
+    // Same for explicit conversions
+    EXPECT_TRUE(testExplicitConversion(int32Type, int32Type));
+    EXPECT_TRUE(testExplicitConversion(float32Type, float32Type));
+    EXPECT_TRUE(testExplicitConversion(int32PtrType, int32PtrType));
+    EXPECT_TRUE(testExplicitConversion(boolType, boolType));
+}
+
+// Test invalid conversions
+TEST_F(ConversionConstraintTest, InvalidConversions)
+{
+    // Integer to float should not be allowed
+    EXPECT_FALSE(testImplicitConversion(int32Type, float32Type));
+    EXPECT_FALSE(testExplicitConversion(int32Type, float32Type));
+    
+    // Float to integer should not be allowed
+    EXPECT_FALSE(testImplicitConversion(float32Type, int32Type));
+    EXPECT_FALSE(testExplicitConversion(float32Type, int32Type));
+    
+    // Integer to bool should not be allowed
+    EXPECT_FALSE(testImplicitConversion(int32Type, boolType));
+    EXPECT_FALSE(testExplicitConversion(int32Type, boolType));
+    
+    // Bool to integer should not be allowed
+    EXPECT_FALSE(testImplicitConversion(boolType, int32Type));
+    EXPECT_FALSE(testExplicitConversion(boolType, int32Type));
+}
+
+// Test that conversion constraints can be created and applied
+TEST_F(ConversionConstraintTest, ConversionConstraintApplication)
```

Comment:
The application tests should test that they are correctly applied, substituted and that the AST Module will be updated to add the cast expressions for implicit casts. Look at the other ConstraintSystem tests.